Test collection hook to ignore uninstallable addons

sbidoul opened the pull request:
This is useful to use pytest (and IDE test discovery) in migration projects where we have a bunch of uninstallable addons.

sbidoul committed:
Test collection hook to ignore tests in uninstallable addons

sbidoul commented:
@baimont @rousseldenis @simahawk

simahawk commented on pytest_odoo.py:
```diff
@@ -186,3 +187,33 @@ def pytest_pycollect_makemodule(path, parent):
+def _find_manifest_path(collection_path: Path) -> Path:
+    """Try to locate an Odoo manifest file in the collection path."""
+    # check if collection_path is an addon directory
+    manifest_path = collection_path / "__manifest__.py"
+    if manifest_path.is_file():
+        return manifest_path
+    # check if collection_path is the tests directory of an addon
+    if collection_path.name == "tests":
+        manifest_path = collection_path.parent / "__manifest__.py"
+        if manifest_path.is_file():
+            return manifest_path
+    # cehck if collection_path is a file in the tests directory of an addon
+    if collection_path.parent.name == "tests":
+        manifest_path = collection_path.parent.parent / "__manifest__.py"
+        if manifest_path.is_file():
+            return manifest_path
```
```suggestion
    candidates = (
        collection_path,  # addon directory
        collection_path.parent,  # tests dir of an addon
        collection_path.parent.parent  # file in the tests dir of an addon
    )
    for path in candidates:
        manifest_path = path / "__manifest__.py"
        if manifest_path.is_file():
            return manifest_path
```

simahawk reviewed (commented):
LG, just code review. Draft because you are still testing?

sbidoul committed:
Don't inject odoo.addons namespace outside of Odoo addons

This is necessary to allow testing regular code while pytest-odoo
is installed.

sbidoul committed:
Search manifest more aggressively

Co-authored-by: Simone Orsi <simahawk@users.noreply.github.com>

sbidoul committed:
Pre-import odoo.tests

Otherwise the load_registry autouse fixture fails if
no module that imports odoo.tests was imported by
the test collector.

sbidoul commented:
@simahawk this should now be ready.

I took your suggestion to search for manifests more aggressively.

I also added here two commits to make pytest-odoo behave better when running non-odoo tests in an environment where pytest-odoo is installed.

petrus-v reviewed (approved):
Code review LGTM
